---
title: T Cell Clonal Analysis using Single-Cell RNA Sequencing and Reference Maps
author: "M. Andreatta <massimo.andreatta at unil.ch> and S. Carmona <santiago.carmona at unil.ch>"
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file,
 encoding=encoding,
 output_file=file.path(dirname(input_file), out_dir, 'index.html'))})
#output: html_notebook
---

# Introduction

In this notebook, we outline a computational workflow to perform T cell states and clonal analysis from scRNA-seq data, based on the R packages Seurat, ProjecTILs and scRepertoire. Given a scRNA-seq T cell dataset with TCR sequence information, cell states are automatically annotated by reference projection using the ProjecTILs method. TCR information is used to track individual clonotypes, assess their clonal expansion, proliferation rates, bias towards specific differentiation states, and the clonal overlap between T cell subtypes.

In this workflow we will analyze a dataset of droplet-based paired scRNA-seq and scTCR-seq (10x Genomics) of tumor-infiltrating T cells from patients with advanced basal cell carcinoma (BCC), from the study by [Yost et al](https://www.ncbi.nlm.nih.gov/pubmed/31359002).
The protocol is adaptable to any single-cell transcriptomics data with TCR sequence information in human and mouse. 


# R environment

Install and load all required packages. 

```{r, message=FALSE, warning=F, results=FALSE}
if (!require("renv")) install.packages("renv")
library(renv)
renv::restore()

library(remotes)
library(BiocManager)

#Specify versions for reproducibility
#remotes::install_github("carmonalab/ProjecTILs", ref="7de06ad")
#remotes::install_github("ncborcherding/scRepertoire", ref="4441960")

library(ggplot2)
library(Seurat)
library(ProjecTILs)
library(scRepertoire)
library(GEOquery)
library(stringr)
library(patchwork)

# Increase timeout limit for slow connections
options(timeout = max(120, getOption("timeout")))
```


# Data pre-processing

Download processed counts matrices, TCR and cells metadata from [GEO entry](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE123813).
Create Seurat object

```{r}
geo_acc <- "GSE123813"
datadir <- "data"

cached_seurat <- sprintf("%s/Yost_alldata_seurat.rds", datadir)

#If we have already processed the data, read in a cached version
if(file.exists(cached_seurat)){
  seurat <- readRDS(cached_seurat)
} else {
  
  dir.create(datadir, showWarnings = FALSE)
  dir.create("plots", showWarnings = FALSE)
  
  gse <- getGEO(geo_acc)
  getGEOSuppFiles(geo_acc, baseDir=datadir)
  
  #Read in expression matrix
  exp.mat <- read.delim(sprintf("%s/%s/GSE123813_bcc_scRNA_counts.txt.gz",
                                datadir, geo_acc), header=T, sep="\t")
  seurat <- CreateSeuratObject(counts = exp.mat, project = "Yost",
                                     min.cells = 3, min.features = 10)
  rm(exp.mat)
  
  #and the metadata associated with the GEX data
  meta <- read.delim(sprintf("%s/%s/GSE123813_bcc_all_metadata.txt.gz",
                             datadir, geo_acc), header = T, sep = "\t")
  meta.T <- read.delim(sprintf("%s/%s/GSE123813_bcc_tcell_metadata.txt.gz",
                               datadir, geo_acc), header = T, sep = "\t")
  
  seurat <- AddMetaData(seurat, meta$patient, col.name="patient")
  
  rownames(meta) <- meta$cell.id

  meta.T.c <- meta.T$cluster
  names(meta.T.c) <- meta.T$cell.id

  seurat <- AddMetaData(seurat, meta)
  seurat <- AddMetaData(seurat, meta.T.c, col.name = "cluster.T")
  
  #Read in TCR-seq data
  tcr.file <- sprintf("%s/%s/GSE123813_bcc_tcr.txt.gz", datadir, geo_acc)
  tcr.data <- read.csv(tcr.file, sep = "\t")
  
  #Check whether there is a alpha-beta pair
  alpha <- str_count(tcr.data$cdr3s_aa, pattern = "TRA:")
  beta <- str_count(tcr.data$cdr3s_aa, pattern = "TRB:")
  
  tcr.data$has_alphabeta <- ifelse(alpha>0 & beta>0, 1, 0)
  tcr.data$has_singlepair <- ifelse(alpha==1 & beta==1, 1, 0)
  
  seurat <- AddMetaData(seurat, tcr.data)
  
  #Make unique IDs for TCRs per patient
  seurat$cdr3s_pat <- NA
  inds <- !is.na(seurat$cdr3s_aa)
  seurat$cdr3s_pat[inds] <- paste0(seurat$cdr3s_aa[inds],
                                         "_",seurat$patient[inds])
  
  #Save object to disk
  saveRDS(seurat, cached_seurat)
}

```


```{r}
head(sort(table(seurat$cdr3s_pat), decreasing = T))
```


Subset on pre-treatment samples
```{r}
table(seurat$treatment, seurat$patient)

seurat.pre <- subset(seurat, subset=treatment=="pre")
```

# ProjecTILs analysis

## Load reference CD8 T map

Here we will project the query data onto a human reference TIL atlas for CD8 T cells, with the goal to interpret T cell diversity in the context of reference T cell subtypes. First, load the reference map:
```{r}
ref.file <- "CD8T_human_ref_v1.rds"
if (!file.exists(ref.file)) {
  download.file("https://figshare.com/ndownloader/files/38921366", destfile = ref.file)
}

ref.cd8 <- load.reference.map("CD8T_human_ref_v1.rds")
```

Let's have a look at the reference CD8 T cell subtypes
```{r}
DimPlot(ref.cd8, cols = ref.cd8@misc$atlas.palette, label = T, raster = T) +
  theme(aspect.ratio = 1) + ggtitle("CD8 T reference") + NoLegend()

ggsave("plots/CD8T_reference_umap.pdf", height = 4, width=5)
```


Show key marker genes for the reference subtypes

```{r fig.height=4, fig.width=3}
DefaultAssay(ref.cd8) <- "RNA"

genes <- c("SELL", "TCF7", "LEF1", "CCR7", "S1PR1", "LMNA", "IL7R", "GZMK", "FGFBP2",
    "FCGR3A", "XCL1", "XCL2", "CD200", "CRTAM", "GNG4", "TOX", "PDCD1", "HAVCR2",
    "GZMB", "PRF1", "LAG3", "KLRB1", "TRAV1-2")

VlnPlot(ref.cd8, features = genes, stack = TRUE, flip = TRUE, fill.by = "ident",
    cols = ref.cd8@misc$atlas.palette) + NoLegend()

ggsave("plots/CD8T_ref_subset_violins.pdf", height=12, width=6)

DefaultAssay(ref.cd8) <- "integrated"
```

## Reference-based analysis

Split object by patient and project them separately into the reference map
```{r}
#Split by patient
seurat.list <- SplitObject(seurat.pre, split.by = "patient")

ncores = 6
seurat.projected <- Run.ProjecTILs(seurat.list, ref.cd8, ncores = ncores)
```



We can verify the expression profile using a panel of marker genes:
```{r fig.height=4, fig.width=8}
genes4radar <- c("CD4", "CD8A", "TCF7", "CCR7", "IL7R", "LMNA", "GZMA", "GZMK", "FGFBP2",
    "XCL1", "CRTAM", "TOX", "PDCD1", "HAVCR2", "PRF1", "GZMB", "TRAV1-2",
    "KLRB1")

plot.states.radar(ref.cd8, query = seurat.projected$su009, genes4radar = genes4radar)
ggsave("plots/radars_projected.pdf", height=11, width=16)

```
Expression profile match very well those of the reference map



What is the subtype composition in different patients? Visualize projection and composition for individual patients:
```{r}
which.patient <- "su009"
a <- plot.projection(ref.cd8, query = seurat.projected[[which.patient]],
                     pointsize = 0.1, linesize = 0.3, raster=T) + NoLegend() +
                    ggtitle(which.patient)
b <- plot.statepred.composition(ref.cd8, query = seurat.projected[[which.patient]]) 
a | b

ggsave("plots/single_patient_composition.pdf", height=4, width=8)
```

Remove samples with very few cells. A minimum number of cells is required for robust analyses.
```{r fig.width=6}
sizes <- as.vector(lapply(seurat.projected, ncol))

# At least 100 cells
keep <- names(sizes)[sizes > 100]

seurat.projected <- seurat.projected[keep]

plots <- lapply(names(seurat.projected), function(x) {
  plot.statepred.composition(ref.cd8, query = seurat.projected[[x]],
                             metric = "Percent") + ggtitle(x)
})
names(plots) <- names(seurat.projected)

#sort by fraction of Tex cells
tex.fraction <- lapply(seurat.projected, function(x) {
  t <- table(x$functional.cluster)
  unname(t["CD8.TEX"]/sum(t))
})
tex.fraction <- unlist(tex.fraction)

plots <- plots[names(sort(tex.fraction, decreasing = T))]

wrap_plots(plots, ncol=4)

ggsave("plots/all_patient_composition.pdf", height=6, width=14)
```

Merge data into a single object to enable analyses across patients
```{r  fig.width=12, fig.height=8}
# First, merge data into single object
merged.projected <- Reduce(merge.Seurat.embeddings, seurat.projected)
Idents(merged.projected) <- "functional.cluster"
DefaultAssay(merged.projected) <- "RNA"
```

# Clonotype analysis

First, calculate frequencies over the total of cells with a TCR chain in the same sample:
```{r fig.width=5, fig.height=3}
freqs <- lapply(seurat.projected, function(x) {
  table(x$cdr3s_pat) / sum(!is.na(x$cdr3s_pat))
})
freqs <- Reduce(c, freqs)
sorted <- sort(freqs, decreasing = TRUE)
largest.clones <- head(sorted, 6)
```

Plot individual clones in the reference space:
```{r fig.width=5, fig.height=3}
plots <- list()
for (i in 1:length(largest.clones)) {
  
  ctype <- names(largest.clones)[i]
  patient <- gsub("\\S+_","", ctype, perl=T)
  
  cells.in.clone <- which(merged.projected[["cdr3s_pat"]]==ctype)
  plots[[i]] <- plot.projection(ref.cd8, merged.projected[,cells.in.clone],
                                pointsize = 1, linesize = 0.3, raster=T) + 
    ggtitle(sprintf("%s - Freq %.1f %%", patient, 100*largest.clones[i])) + NoLegend()
}

wrap_plots(plots, ncol = 3)
ggsave("plots/largest_clones_umap.pdf", height=7, width=12)
```

## Clonal expansion

Summarize clonal sizes:
```{r}
cutoffs <- c(0, 1, 5, 20, 50)
names(cutoffs) <- c("Single (=1)","Small (<=5)","Medium (<=20)","Large (<=50)","Hyper (>50)")

call <- "cdr3s_pat"
tab <- table(merged.projected[[call]])
merged.projected$clonesize <- NA
merged.projected$cloneType <- NA

for (r in 1:nrow(merged.projected[[]])) {
  e <- merged.projected@meta.data[r,call]
  if (!is.na(e) & e %in% names(tab)) {
    merged.projected@meta.data[r,"clonesize"] <- tab[e]
  }
}

for (i in seq_along(cutoffs)) {
  merged.projected$cloneType[merged.projected$clonesize > cutoffs[i]] <- names(cutoffs[i])
}
merged.projected$cloneType <- factor(merged.projected$cloneType, levels= rev(names(cutoffs)))

merged.projected$Frequency <- merged.projected$clonesize

#table(merged.projected$cloneType, merged.projected$clonesize)
```


See clonal composition and clonal diversity per cell subtype:
```{r}
combined <- expression2List(merged.projected, split.by = "functional.cluster")

clonalDiversity(combined, cloneCall = "cdr3s_pat") +
  scale_color_manual(values = ref.cd8@misc$atlas.palette)

clonalProportion(combined, cloneCall = "cdr3s_pat", split = c(5, 10, 50, 100, 500, 2000))
  
clonalHomeostasis(combined, cloneCall = "cdr3s_pat",
                  cloneTypes = c(Rare=0.001, Small = 0.005, Medium = 0.01, Large = 0.1, Hyperexpanded= 1))

compareClonotypes(combined, 
                  clonotypes = names(largest.clones),
                  cloneCall="cdr3s_pat", 
                  graph = "alluvial")
```

A useful visualization is the occupied single-cell repertoire. It measures the number of cells, and their clonal expansion type, for each of the cell subtypes:
```{r}
occupiedscRepertoire(merged.projected, x.axis = "functional.cluster") + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave("plots/occupiedRep_barplot.pdf", height=5, width=6)  
```

## Clonotype proliferation rate

Identify most proliferating clones based on cell cycling signatures. If ProjecTILs was run with `filter.cells=TRUE` (which is the default), cell cycling scores are already pre-calculated and available as metadata:

```{r}
#Identify cycling cells
merged.projected$is.cycling <- ifelse((merged.projected$cycling.score.G1_S > 0.1 |
                                 merged.projected$cycling.score.G2_M > 0.1),
                                 yes = "Proliferating",
                                 no = "Resting")

clonotypes <- table(merged.projected$cdr3s_pat)
expanded <- names(clonotypes)[clonotypes>=2]

prolif.freq <- sapply(expanded, function(x) {
  sub <- subset(merged.projected[[]], subset=cdr3s_pat == x)
  sum(sub$is.cycling == "Proliferating") / ncol(sub)
})

sorted <- sort(prolif.freq, decreasing = T)
frequency.proliferating <- head(sorted, 6)
```


See frequency of top 6 clones
```{r fig.height=1.5, fig.width=4}
df <- data.frame(frequency.proliferating)
df$clone <- rownames(df)
ggplot(df, aes(x=reorder(clone, frequency.proliferating), y=frequency.proliferating)) +
  geom_bar(stat="identity", fill="black") + coord_flip() +
  theme_light()
ggsave("plots/most_prolif_clones_barplot.pdf", height=3, width=10)
```

See them on the reference space
```{r fig.width=5, fig.height=3}
plots <- list()
for (ctype in names(frequency.proliferating)) {
  
  freq <- frequency.proliferating[ctype]
  title <- sprintf("%.0f %% proliferating", 100*freq)
  cells.in.clone <- which(merged.projected[["cdr3s_pat"]]==ctype)
  
  plots[[ctype]] <- plot.projection(ref.cd8, merged.projected[,cells.in.clone],
                                pointsize = 1, linesize = 0.3, raster=T) + 
    ggtitle(title) + NoLegend()
}

wrap_plots(plots, ncol = 3)
ggsave("plots/most_prolif_clones_umap.pdf", height=7, width=12)
```

## Clonal sharing

Metrics of clonal overlap (e.g. Morisita index) can be used to assess clonal sharing between subtypes:
```{r}
clonalOverlap(combined, cloneCall = "cdr3s_pat", method = "morisita") + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
ggsave("plots/morisita_heatmap.pdf", height=4, width=6)
```

Subset connectivity using circos plots
```{r, fig.width=6}
library(circlize)

circles <- getCirclize(merged.projected,
                       cloneCall = "cdr3s_pat",
                       group.by = "functional.cluster")

#Graphing the chord diagram
pdf("plots/circos_subtype.pdf", width = 5, height = 5)
circlize::chordDiagram(circles,
                       self.link = 1, 
                       grid.col = ref.cd8@misc$atlas.palette)
dev.off()
```

Only for large clones (e.g. with 5 cells or more):
```{r, fig.width=8}
large <- subset(merged.projected, subset = clonesize >= 5)

circles <- getCirclize(large,
                       cloneCall = "cdr3s_pat",
                       group.by = "functional.cluster")

#Graphing the chord diagram
circlize::chordDiagram(circles,
                       self.link = 1, 
                       grid.col = ref.cd8@misc$atlas.palette)

```

## Clonotype bias

Find clones with significant enrichment for specific cell subtypes:
```{r}
clonotypeBias(merged.projected, cloneCall = "cdr3s_pat", split.by = "patient",
              group.by = "functional.cluster", min.expand = 10, n.boots = 20) +
  scale_color_manual(values = ref.cd8@misc$atlas.palette)

ggsave("plots/clonotype_bias.pdf", height=3, width=5)

biased <- clonotypeBias(large, cloneCall = "cdr3s_pat", split.by = "patient",
              group.by = "functional.cluster", min.expand = 10, exportTable = TRUE)

most.biased <- biased[order(biased$Z.score, decreasing = TRUE),]
head(most.biased)
```

See where the most "biased" clonotypes are found on the reference.
```{r fig.width=5, fig.height=3}
plots <- list()
for (i in 1:6) {
  
  ctype <- most.biased[i, "Clone"]
  patient <- most.biased[i, "Sample"]
  size <- most.biased[i, "ncells"]
  zscore <- most.biased[i, "Z.score"]
  cells.in.clone <- which(merged.projected[["cdr3s_pat"]]==ctype)
  title <- sprintf("Clone %s, size %s, Z=%.1f", i, size, zscore)
  
  plots[[i]] <- plot.projection(ref.cd8, merged.projected[,cells.in.clone],
                                pointsize = 1, linesize = 0.3, raster=T) + 
    ggtitle(title) + NoLegend()
}

wrap_plots(plots, ncol = 3)

ggsave("plots/biased_clones_umap.pdf", height=7, width=12)
```
